### PR TITLE
Rewrite Ollama chat handler to support proper streaming responses

### LINKED ISSUE #1938: [Bug]: Fabric does connect to Ollama on CLI, but not with OpenWebUI
### What happened?

Hi, 
I'm running fabric in a docker-compose environment. Using the CLI, it seems to work, so, Ollama (running on an other server) get's to work, and provide feedback.
Like this:
``` sh
> docker run --rm -it -v $HOME/.fabric-config:/root/.config/fabric kayvan/fabric:latest -p summarize
ONE SENTENCE SUMMARY: No content provided for summarization.

MAIN POINTS:
1. No points available.

TAKEAWAYS:
1. No takeaways available.
```

Then I've connected Fabric to my OpenWebUI, the models are being displayed, but if I try to use them, I get an error 500.
This is my docker-compose.yml:
```
---
services:
  fabric:
    image: kayvan/fabric:latest
    container_name: fabric
#    privileged: true
# Der Befehl, der beim Start ausgeführt wird
    command: [
      "--serveOllama",
      "--debug=3",
#      "--serve", #removed, as I dont need the fabric API
      "--api-key",   "${FABRIC_API_KEY}"
      ]
    environment:
#      - PUID=1000
#      - PGID=1000
      - TZ=Europe/Berlin
#      - GIN_TRUSTED_PROXIES=10.0.0.0/8,192.168.178.0/24
    volumes:
      - ${HOME}/.fabric-config:/root/.config/fabric

    # Port‑Mapping: 8080 (Host) ↔ 8080 (Container)
    ports:
      - "8180:8080"
    restart: unless-stopped```


### Operating System

Linux - amd64

### OS Version

```shell
PRETTY_NAME="Debian GNU/Linux 13 (trixie)"
NAME="Debian GNU/Linux"
VERSION_ID="13"
VERSION="13 (trixie)"
VERSION_CODENAME=trixie
DEBIAN_VERSION_FULL=13.3
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

### How did you install Fabric?

Other (`docker-compose`)

### Version

```text
# docker run --rm -it -v $HOME/.fabric-config:/root/.config/fabric kayvan/fabric:latest --version
v1.4.380
```

### Relevant log output

```shell
fabric  | [GIN] 2026/01/16 - 13:21:40 | 200 |       8.947µs | 192.168.178.137 | GET      "/api/version"
fabric  | [GIN] 2026/01/16 - 13:22:05 | 500 |     268.678µs | 192.168.178.137 | POST     "/api/chat"
fabric  | 2026/01/16 13:22:05 Error marshalling body: json: unsupported type: chan domain.StreamUpdate
```

### Relevant screenshots (optional)

The list of patterns are being displayed in OpenWebUI:

Image

My chat with one of the patterns:

Image

It doesn't matter, which pattern I use, it's always an Error 500

### PR DESCRIPTION
**Pull request by ksylvan**
# Rewrite Ollama chat handler to support proper streaming responses

## Summary

This pull request refactors the Ollama chat handler in the Fabric REST API to support proper streaming responses, replacing the previous single-read body parsing with a streaming `bufio.Scanner` approach. It also improves error handling, adds support for Fabric-specific variables, and aligns duration fields to int64 nanosecond precision for consistency. Additionally, it includes unit tests for the new URL building function and updates the domain struct to exclude a channel from JSON serialization.

## Related Issues

Closes #1938 

## Files Changed

- **cmd/generate_changelog/incoming/1940.txt** (new file): Adds a changelog entry documenting the changes in this PR, including the refactoring details and improvements.
- **internal/domain/domain.go** (modified): Updates the `ChatOptions` struct to add a `json:"-"` tag to the `UpdateChan` field, preventing it from being serialized in JSON responses.
- **internal/server/ollama.go** (modified): Major refactoring of the `ollamaChat` function and related code to support streaming SSE responses, improve error handling, and add new helper functions for URL building and response writing.
- **internal/server/ollama_test.go** (new file): Introduces comprehensive unit tests for the `buildFabricChatURL` function, covering various input scenarios including valid URLs, malformed inputs, and edge cases.

### internal/domain/domain.go
- Added `json:"-"` tag to `UpdateChan chan StreamUpdate` in the `ChatOptions` struct. This ensures the channel is not included in JSON serialization, which is appropriate since channels cannot be marshaled to JSON and could cause issues in API responses.

### internal/server/ollama.go
- **Struct Updates**: Modified `OllamaRequestBody` to include `Options map[string]any` and `Variables map[string]string` for better handling of Fabric-specific variables. Updated duration fields in `OllamaResponse` from `int` to `int64` for nanosecond precision consistency.
- **Variable Extraction**: Added logic to extract variables from either the top-level `Variables` field or `Options.variables`, with support for JSON string parsing or direct map conversion. This enhances flexibility in handling user-provided variables.
- **URL Building**: Replaced direct URL construction with a new `buildFabricChatURL` function that validates and normalizes addresses, supporting HTTP/HTTPS URLs, port-only shorthands, and bare host:port formats while rejecting invalid inputs like paths in bare addresses.
- **Streaming Implementation**: Switched from `io.ReadAll` to `bufio.Scanner` for processing SSE streams line-by-line. Added streaming response writing with `writeOllamaResponse` and `writeOllamaResponseStruct` functions, including proper flushing for real-time output.
- **Error Handling Improvements**: Replaced `log.Fatal` with proper HTTP error responses to prevent server crashes. Added upstream error propagation, including status code checks and error message forwarding in both streaming and non-streaming modes.
- **Response Building**: Introduced `buildFinalOllamaResponse` helper to construct consistent final responses with accurate timing metrics.
- **Context Propagation**: Updated to use `c.Request.Context()` instead of a background context for better request lifecycle management.

Key code snippet for streaming logic:
```go
scanner := bufio.NewScanner(fabricRes.Body)
scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
for scanner.Scan() {
    line := scanner.Text()
    if !strings.HasPrefix(line, "data: ") {
        continue
    }
    // Parse and handle each SSE payload
}
```

### internal/server/ollama_test.go
- Added unit tests for `buildFabricChatURL` with 12 test cases covering valid inputs (e.g., full URLs, port shorthands) and invalid ones (e.g., empty strings, malformed URLs, paths in bare addresses). This ensures robust URL validation.

## Reason for Changes

The original Ollama chat handler used a blocking `io.ReadAll` approach that read the entire response body at once, which was inefficient for large or streaming responses and could lead to memory issues. It also lacked proper error handling, potentially crashing the server with `log.Fatal`, and did not support real-time streaming as expected by Ollama-compatible clients. The changes address these by implementing true SSE streaming, improving reliability, and adding support for Fabric-specific features like variables. The domain change prevents JSON serialization errors, and the tests ensure the new URL building logic is reliable.

## Impact of Changes

- **Performance**: Streaming reduces memory usage for large responses by processing data incrementally rather than buffering everything. This improves scalability for long conversations.
- **Functionality**: Enables real-time streaming responses, making the API more compatible with Ollama clients expecting SSE. Adds variable support for dynamic prompt customization.
- **Reliability**: Better error handling prevents crashes and provides clearer error messages to clients. Duration fields now use int64 for precise timing, avoiding potential overflow issues.
- **Compatibility**: Changes maintain backward compatibility for non-streaming requests while adding streaming capabilities. The `json:"-"` tag avoids serialization issues in existing code paths.
- **Potential Bugs**: 
  - If upstream Fabric responses contain extremely long lines (>1MB), the scanner buffer could fail, triggering an error (handled gracefully). 
  - Variable parsing assumes string values; non-string types in `Options.variables` are silently ignored, which might confuse users if they expect broader type support.
  - Context propagation relies on Gin's request context; if the request is canceled mid-stream, it may not cleanly terminate all goroutines (though HTTP client respects context).

## Test Plan

- **Unit Tests**: The new `ollama_test.go` file includes comprehensive tests for `buildFabricChatURL`, covering all documented input formats and error cases. Run with `go test ./internal/server` to verify URL building logic.
- **Integration Testing**: Manually test streaming and non-streaming Ollama chat requests against a running Fabric server. Verify SSE output with tools like `curl --no-buffer` for real-time streaming. Check error scenarios by simulating upstream failures.
- **Regression Testing**: Ensure existing non-streaming functionality still works by running existing API tests. Validate that duration fields are correctly reported as int64 without breaking clients expecting integers (though this is a minor type change, clients should handle it).
- **Edge Cases**: Test with large responses to confirm streaming doesn't buffer excessively. Verify variable extraction with JSON strings and maps.

## Additional Notes

This refactoring aligns the Ollama handler with modern streaming expectations, reducing latency for interactive use cases. The changelog entry ensures traceability. If deploying, monitor for increased CPU usage due to line-by-line processing, though it's generally more efficient than full buffering. No breaking changes for existing non-streaming clients, but streaming clients will now receive properly formatted NDJSON responses.

### CODE REVIEW

File: internal/server/ollama.go [hunk 1]
```diff
@@ -192,11 +194,13 @@ func (f APIConvert) ollamaChat(c *gin.Context) {
...
-		req, err = http.NewRequest("POST", fmt.Sprintf("http://127.0.0.1%s/chat", *f.addr), bytes.NewBuffer(fabricChatReq))
+	baseURL, err := buildFabricChatURL(*f.addr)
+	if err != nil {
+		log.Printf("Error building /chat URL: %v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err})
+		return
 	}
+	req, err = http.NewRequest("POST", fmt.Sprintf("%s/chat", baseURL), bytes.NewBuffer(fabricChatReq))
 	if err != nil {
 		log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (ksylvan) [reply internal/server/ollama.go]: Fixed.

**Comment by ksylvan** [comment]
## Copilot Review Response

Addressing the remaining 10 review comments:

### 1. Placeholder token counts (lines 310, 333)
**Status: Fixed**
- Removed `PromptEvalCount` and `EvalCount` hardcoded values
- These fields now rely on `omitempty` JSON tags and are excluded from responses
- Token counts should come from upstream models, not be fabricated

### 2. Content-Type header timing (line 238)
**Status: Fixed**
- Moved `Content-Type: application/x-ndjson` header to line 221
- Now set immediately after response validation, before any potential streaming writes
- Ensures all streaming responses (including errors) have correct header

### 3. Struct initialization pattern (lines 302, 325, 367)
**Status: Won't Fix**
- Research shows this is a known Go language limitation (golang/go#29438)
- The double-cast pattern is verbose but currently required for anonymous struct fields
- Alternative (defining named types) would add more complexity for one-off response structures
- Will revisit if Go language evolves to support simpler syntax

### 4. URL validation in buildFabricChatURL (line 348)
**Status: TODO**
- Will add validation to ensure parsed.Host is non-empty
- Prevents malformed URLs like "http://" or "https://:8080"

### 5. Unit tests for buildFabricChatURL (line 355)
**Status: TODO**
- Will add unit tests covering edge cases (empty strings, malformed URLs, different formats)

### 6. Unit tests for streaming logic (line 280)
**Status: TODO**
- Will add tests for streaming vs non-streaming modes, error handling, partial responses

Working on remaining TODOs now.

File: internal/server/ollama.go [hunk 2]
```diff
@@ -190,81 +191,226 @@ func (f APIConvert) ollamaChat(c *gin.Context) {
...
+			return "", fmt.Errorf("invalid address: scheme must be http or https")
+		}
+		if parsed.Host == "" {
+			return "", fmt.Errorf("invalid address: missing host")
 		}
-		res = append(res, marshalled...)
-		res = append(res, '\n')
+		if strings.HasPrefix(parsed.Host, ":") {
+			return "", fmt.Errorf("invalid address: missing hostname")
+		}
```
[the remark this answers is not in the document]
Comment (ksylvan) [reply internal/server/ollama.go:387]: This check is needed to catch port-only URLs.

File: internal/server/ollama.go [hunk 3]
```diff
@@ -190,81 +191,226 @@ func (f APIConvert) ollamaChat(c *gin.Context) {
...
+			return
+		}
+		if fabricResponse.Type == "error" {
+			if prompt.Stream {
+				// In streaming mode, propagate the upstream error via a final streaming chunk
+				_ = writeOllamaResponse(c, prompt.Model, fmt.Sprintf("Error: %s", fabricResponse.Content), true)
+			} else {
+				c.JSON(http.StatusInternalServerError, gin.H{"error": fabricResponse.Content})
+			}
+			return
```
[the remark this answers is not in the document]
Comment (ksylvan) [reply internal/server/ollama.go:291]: This is the correct implementation according to the Official Ollama docs.

File: internal/server/ollama.go [hunk 4]
```diff
@@ -190,81 +191,226 @@ func (f APIConvert) ollamaChat(c *gin.Context) {
...
+		}
+		if fabricResponse.Type != "content" {
+			continue
+		}
+		contentBuilder.WriteString(fabricResponse.Content)
+		if prompt.Stream {
+			if err := writeOllamaResponse(c, prompt.Model, fabricResponse.Content, false); err != nil {
+				log.Printf("Error writing response: %v", err)
+				// Attempt to send a final error message to properly close the stream
+				_ = writeOllamaResponse(c, prompt.Model, "Error: failed to write response", true)
```
[the remark this answers is not in the document]
Comment (ksylvan) [reply internal/server/ollama.go]: Fixed

File: internal/server/ollama.go [hunk 5]
```diff
@@ -190,81 +191,226 @@ func (f APIConvert) ollamaChat(c *gin.Context) {
...
+	if err != nil {
+		return err
+	}
+	if _, err := c.Writer.Write(marshalled); err != nil {
+		return err
+	}
+	if _, err := c.Writer.Write([]byte("\n")); err != nil {
+		return err
+	}
+	c.Writer.Flush()
```
[the remark this answers is not in the document]
Comment (ksylvan) [reply internal/server/ollama.go]: Fixed

File: internal/server/ollama.go [hunk 6]
```diff
@@ -190,81 +191,226 @@ func (f APIConvert) ollamaChat(c *gin.Context) {
...
+		var fabricResponse FabricResponseFormat
+		if err := json.Unmarshal([]byte(payload), &fabricResponse); err != nil {
+			log.Printf("Error unmarshalling body: %v", err)
+			if prompt.Stream {
+				// In streaming mode, send the error in the same streaming format
+				_ = writeOllamaResponse(c, prompt.Model, "Error: failed to parse upstream response", true)
+			} else {
+				c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to unmarshal Fabric response"})
+			}
+			return
```
[the remark this answers is not in the document]
Comment (ksylvan) [reply internal/server/ollama.go:282]: As before, this is the correct implementation.

File: internal/server/ollama_test.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+			if (err != nil) != tt.wantErr {
+				t.Errorf("buildFabricChatURL() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("buildFabricChatURL() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (ksylvan) [reply internal/server/ollama_test.go:100]: Thanks. May follow up with other PRs later.
